### PR TITLE
[aptos-move-package] Sort modules in packages by name

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
This allows users to easily tell which order they need to upload their modules in a package, and helps readability.  A flag is used to turn on this new feature for backwards compatibility when verifying packages.

### Test Plan
```
 ./target/debug/aptos move compile --save-metadata --package-dir aptos-move/framework/aptos-framework --named-addresses hello_blockchain=default
Compiling, may take a little while to download git dependencies...
INCLUDING DEPENDENCY AptosStdlib
INCLUDING DEPENDENCY MoveStdlib
BUILDING AptosFramework
{
  "Result": [
    "0000000000000000000000000000000000000000000000000000000000000001::account",
    "0000000000000000000000000000000000000000000000000000000000000001::aggregator",
    "0000000000000000000000000000000000000000000000000000000000000001::aggregator_factory",
    "0000000000000000000000000000000000000000000000000000000000000001::aptos_account",
    "0000000000000000000000000000000000000000000000000000000000000001::aptos_coin",
    "0000000000000000000000000000000000000000000000000000000000000001::aptos_governance",
    "0000000000000000000000000000000000000000000000000000000000000001::block",
    "0000000000000000000000000000000000000000000000000000000000000001::chain_id",
    "0000000000000000000000000000000000000000000000000000000000000001::chain_status",
    "0000000000000000000000000000000000000000000000000000000000000001::code",
    "0000000000000000000000000000000000000000000000000000000000000001::coin",
    "0000000000000000000000000000000000000000000000000000000000000001::consensus_config",
    "0000000000000000000000000000000000000000000000000000000000000001::event",
    "0000000000000000000000000000000000000000000000000000000000000001::gas_schedule",
    "0000000000000000000000000000000000000000000000000000000000000001::genesis",
    "0000000000000000000000000000000000000000000000000000000000000001::governance_proposal",
    "0000000000000000000000000000000000000000000000000000000000000001::guid",
    "0000000000000000000000000000000000000000000000000000000000000001::managed_coin",
    "0000000000000000000000000000000000000000000000000000000000000001::optional_aggregator",
    "0000000000000000000000000000000000000000000000000000000000000001::reconfiguration",
    "0000000000000000000000000000000000000000000000000000000000000001::resource_account",
    "0000000000000000000000000000000000000000000000000000000000000001::stake",
    "0000000000000000000000000000000000000000000000000000000000000001::staking_config",
    "0000000000000000000000000000000000000000000000000000000000000001::staking_contract",
    "0000000000000000000000000000000000000000000000000000000000000001::staking_proxy",
    "0000000000000000000000000000000000000000000000000000000000000001::state_storage",
    "0000000000000000000000000000000000000000000000000000000000000001::storage_gas",
    "0000000000000000000000000000000000000000000000000000000000000001::system_addresses",
    "0000000000000000000000000000000000000000000000000000000000000001::timestamp",
    "0000000000000000000000000000000000000000000000000000000000000001::transaction_context",
    "0000000000000000000000000000000000000000000000000000000000000001::transaction_fee",
    "0000000000000000000000000000000000000000000000000000000000000001::transaction_validation",
    "0000000000000000000000000000000000000000000000000000000000000001::util",
    "0000000000000000000000000000000000000000000000000000000000000001::version",
    "0000000000000000000000000000000000000000000000000000000000000001::vesting",
    "0000000000000000000000000000000000000000000000000000000000000001::voting"
  ]
}
```

I also manually verified the BCS metadata file to have the ordered modules

### CODE REVIEW

**Comment by wrwg** [comment]
I'm really surprised that there is a per-name required order. If one I would have expected dep order.

File: aptos-move/e2e-move-tests/src/harness.rs
Comment (movekevin) [inline aptos-move/e2e-move-tests/src/harness.rs:277]: What is this for?

**Review by movekevin — APPROVED** [review]
This makes sense high-level wise. We just need to decide whether we want to sort alphabetically or by dependencies.

**Comment by movekevin** [comment]
> I'm really surprised that there is a per-name required order. If one I would have expected dep order.

I think alphabetic is likely more stable in case the code changes. Dep order makes sense on the technical side but can be less human-readable as well (e.g. If I need to look and find where a module is in the package)

**Review by davidiw — COMMENTED** [review]
@wrwg, if someone wants to build an "Aptos Compiler Server" and publish using an SDK, they must package the module bytecode such that it is ordered per the metadata. Unfortunately, the current ordering is not useful as it cannot be determined given the filenames. This PR should unblock this partner

@gregnazario , before landing can you confirm that this:
* Works
* Doesn't result in incompatible changes (as in I can upgrade a module with this from unsorted to sorted)

File: aptos-move/e2e-move-tests/src/harness.rs
Comment (gregnazario) [reply aptos-move/e2e-move-tests/src/harness.rs:277]: I chose to be sorted in the tests

**Comment by gregnazario** [comment]
> @wrwg, if someone wants to build an "Aptos Compiler Server" and publish using an SDK, they must package the module bytecode such that it is ordered per the metadata. Unfortunately, the current ordering is not useful as it cannot be determined given the filenames. This PR should unblock this partner
> 
> @gregnazario , before landing can you confirm that this:
> 
> * Works
> * Doesn't result in incompatible changes (as in I can upgrade a module with this from unsorted to sorted)

Yep, I'm going to test this out today

**Comment by wrwg** [comment]
> > @wrwg, if someone wants to build an "Aptos Compiler Server" and publish using an SDK, they must package the module bytecode such that it is ordered per the metadata. Unfortunately, the current ordering is not useful as it cannot be determined given the filenames. This PR should unblock this partner
> > @gregnazario , before landing can you confirm that this:
> > 
> > * Works
> > * Doesn't result in incompatible changes (as in I can upgrade a module with this from unsorted to sorted)
> 
> Yep, I'm going to test this out today

Why would this order be needed? I just checked `aptos_framework::code` both Move and native and do not see a reason for this. Is this a bug? Or does the partner need this ordering? 

Also 'ordered per metadata' does not match what you said in the PR description: 

> This allows users to easily tell which order they need to upload their modules in a package, and helps readability. A flag is used to turn on this new feature for backwards compatibility when verifying packages.

There is no order in which users upload modules.

Apologies this PR is rather confusing (bug? feature?) and as I wrote all of this code, I try to understand what do you want to achieve here.

**Comment by davidiw** [comment]
> > > @wrwg, if someone wants to build an "Aptos Compiler Server" and publish using an SDK, they must package the module bytecode such that it is ordered per the metadata. Unfortunately, the current ordering is not useful as it cannot be determined given the filenames. This PR should unblock this partner
> > > @gregnazario , before landing can you confirm that this:
> > > 
> > > * Works
> > > * Doesn't result in incompatible changes (as in I can upgrade a module with this from unsorted to sorted)
> > 
> > 
> > Yep, I'm going to test this out today
> 
> Why would this order be needed? I just checked `aptos_framework::code` both Move and native and do not see a reason for this. Is this a bug? Or does the partner need this ordering?
> 
> Also 'ordered per metadata' does not match what you said in the PR description:
> 
> > This allows users to easily tell which order they need to upload their modules in a package, and helps readability. A flag is used to turn on this new feature for backwards compatibility when verifying packages.
> 
> There is no order in which users upload modules.
> 
> Apologies this PR is rather confusing (bug? feature?) and as I wrote all of this code, I try to understand what do you want to achieve here.

Fair enough, let's get a repro.

**Comment by davidiw** [comment]
It appears to be more on the verification side. I'll let Greg chime in here, but there's no issue on the publishing side. Effectively how does one map the ordering of the modules within code to the metadata, so that they can verify the metadata maps to the code.

**Comment by wrwg** [comment]
> It appears to be more on the verification side. I'll let Greg chime in here, but there's no issue on the publishing side. Effectively how does one map the ordering of the modules within code to the metadata, so that they can verify the metadata maps to the code.

Yes, I checked that code as well. Everything about package metadata verification is our code. Everything about bytecode verification is 100% independent of package metadata. I don't know about any order dependency there (and thats all code I've written), so I suspect a bug, but I've not seen any bug description.

**Comment by wrwg** [comment]
> > It appears to be more on the verification side. I'll let Greg chime in here, but there's no issue on the publishing side. Effectively how does one map the ordering of the modules within code to the metadata, so that they can verify the metadata maps to the code.
> 
> Yes, I checked that code as well. Everything about package metadata verification is our code. Everything about bytecode verification is 100% independent of package metadata. I don't know about any order dependency there (and thats all code I've written), so I suspect a bug, but I've not seen any bug description.

And again, I know about one order dependency (as mentioned above), and this the modules in the `vector<vector<u8>>` code bundle be in bottom up dependency order. This is independent of the metdata order, and is not the alphabetical order either. Perhaps you are trying to guess this order for the metadata, which accidentally (as the compiler produces it) reflects that order? But in general, there is no guarantee right now that metadata gives you the right module order. (And maybe all testing here is just by chance?)

**Comment by gregnazario** [comment]
> > > It appears to be more on the verification side. I'll let Greg chime in here, but there's no issue on the publishing side. Effectively how does one map the ordering of the modules within code to the metadata, so that they can verify the metadata maps to the code.
> > 
> > 
> > Yes, I checked that code as well. Everything about package metadata verification is our code. Everything about bytecode verification is 100% independent of package metadata. I don't know about any order dependency there (and thats all code I've written), so I suspect a bug, but I've not seen any bug description.
> 
> And again, I know about one order dependency (as mentioned above), and this the modules in the `vector<vector<u8>>` code bundle be in bottom up dependency order. This is independent of the metdata order, and is not the alphabetical order either. Perhaps you are trying to guess this order from the metadata, which accidentally (as the compiler produces it) reflects that order? But in general, there is no guarantee right now that metadata gives you the right module order. (And maybe all testing here is just by chance?)

Yeah the ordering is only in the `aptos move verify-package` section, and that's because I check the packages and the order they come in, which might not be important.

After talking to the requester of this, this is more of a ease of readability thing

**Comment by wrwg** [comment]
> > > > It appears to be more on the verification side. I'll let Greg chime in here, but there's no issue on the publishing side. Effectively how does one map the ordering of the modules within code to the metadata, so that they can verify the metadata maps to the code.
> > > 
> > > 
> > > Yes, I checked that code as well. Everything about package metadata verification is our code. Everything about bytecode verification is 100% independent of package metadata. I don't know about any order dependency there (and thats all code I've written), so I suspect a bug, but I've not seen any bug description.
> > 
> > 
> > And again, I know about one order dependency (as mentioned above), and this the modules in the `vector<vector<u8>>` code bundle be in bottom up dependency order. This is independent of the metdata order, and is not the alphabetical order either. Perhaps you are trying to guess this order from the metadata, which accidentally (as the compiler produces it) reflects that order? But in general, there is no guarantee right now that metadata gives you the right module order. (And maybe all testing here is just by chance?)
> 
> Yeah the ordering is only in the `aptos move verify-package` section, and that's because I check the packages and the order they come in, which might not be important.
> 
> After talking to the requester of this, this is more of a ease of readability thing

Just to clarify the alphabetical order is not the order in which the modules need to appear in the `vector<vector<u8>>`. They need to be in definition-use order. If you want to give people a hint of the order you could rely on the order of the module list in metadata, which in turn is derived from the topological order the package system determined(https://github.com/aptos-labs/aptos-core/blob/f279d68ba1f69136166fc2e6789b98a4fc9e8949/aptos-move/framework/src/built_package.rs#L275)

**Pull request closed without merging**
